### PR TITLE
STM32 HRNG driver

### PR DESCRIPTION
**Pull request by networkfusion**
## Description
- Most STM32's support a hardware random number generator. This PR enables this functionality

## Motivation and Context
- Allows more random numbers with less CPU power.
- Makes nanoFramework/Home#123 possible now. 

## How Has This Been Tested?
On an STM32F769I (Be aware, this is my first driver so it needs careful review against the reference implementation.

## Types of changes

- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/include/stm32_rng/hal_stm32_rng.h [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  void rngReleaseModule(void);
+#endif
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif /* HAL_USE_STM32_RNG */
+
+#endif /* HAL_STM32_RNG_H_ */
```
Comment (josesimoes) [inline targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/include/stm32_rng/hal_stm32_rng.h]: missing empty line

File: targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/README.md [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# RNG v1
+
+This driver supports all STM32 series.
```
Comment (josesimoes) [inline targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/README.md]: this is contradicted bellow... something like all but F0 seems more accurate...

File: targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/rng_lld.c [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+    /* Get a 32bit Random number */
+    RNGD1.RandomNumber = RNGD1.Instance->DR;
+
+  return RNGD1.RandomNumber;
+}
+
+
+uint32_t rng_lld_GetLastRandomNumber() {
+return RNGD1.RandomNumber;
```
Comment (josesimoes) [inline targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/rng_lld.c]: missing tab

File: targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/rng_lld.h [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+#if (HAL_USE_STM32_RNG == TRUE)
+
+/*===========================================================================*/
+/* Driver constants.                                                         */
+/*===========================================================================*/
+
+
+// RNG Interrupt definitions
+#define RNG_IT_DRDY  RNG_SR_DRDY  /*!< Data Ready interrupt  */
```
Comment (josesimoes) [inline targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/rng_lld.h]: RNG_SR_DRDY doesn't seem to be used anywhere

File: targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/rng_lld.h [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+#define RNG_IT_CEI   RNG_SR_CEIS  /*!< Clock error interrupt */
+#define RNG_IT_SEI   RNG_SR_SEIS  /*!< Seed error interrupt  */
+
+// RNG Flag definitions
+#define RNG_FLAG_DRDY   RNG_SR_DRDY  /*!< Data ready                 */
```
Comment (josesimoes) [inline targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/rng_lld.h]: RNG_SR_DRDY doesn't seem to be used anywhere

File: targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/rng_lld.h [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+#define RNG_FLAG_CECS   RNG_SR_CECS  /*!< Clock error current status */
```
Comment (josesimoes) [inline targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/rng_lld.h]: this one too

File: targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/rng_lld.h [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+#define RNG_FLAG_SECS   RNG_SR_SECS  /*!< Seed error current status  */
```
Comment (josesimoes) [inline targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/rng_lld.h]: and this one too

File: targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/rng_lld.h [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+// From STMicroelectronics Cube HAL 
+/////////////////////////////////////////////////////////////
+#define rccEnableRNG(lp) rccEnableAHB2(RCC_AHB2ENR_RNGEN, lp)
+#define rccDisableCRC() rccDisableAHB2(RCC_AHB1ENR_CRCEN)
+
+/*===========================================================================*/
+/* Driver macros.                                                            */
+/*===========================================================================*/
+// Reset RNG handle state
+//#define __RNG_RESET_HANDLE_STATE(__HANDLE__) ((__HANDLE__).State = HAL_RNG_STATE_RESET)
```
Comment (josesimoes) [inline targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/rng_lld.h]: can be removed?

File: targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/src/stm32_rng/hal_stm32_rng.c [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+uint32_t rngGetLastRandomNumber() {
+  uint32_t randomNumber;
+  osalSysLock();
+  randomNumber = rng_lld_GetLastRandomNumber();
+  osalSysUnlock();
+  return randomNumber;
+}
+
+#endif /* HAL_USE_STM32_RNG */
```
Comment (josesimoes) [inline targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/src/stm32_rng/hal_stm32_rng.c]: missing empty line

File: targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/rng_lld.h [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+#include "stm32_registry.h"
+
+#if (HAL_USE_STM32_RNG == TRUE)
+
+/*===========================================================================*/
+/* Driver constants.                                                         */
+/*===========================================================================*/
+
+
+// RNG Interrupt definitions
```
Comment (josesimoes) [inline targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/rng_lld.h]: this doesn't seem to be using interrupts so this block can go away, isn't it?

**Comment by josesimoes** [comment]
One last request: on the halconf_nf.h (booter and CLR) for the F091 you have to add that negative RNG driver or the build will break

File: targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/rng_lld.h (hunk 5 above)
Comment (networkfusion) [reply targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/rng_lld.h]: That is used...

File: targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/rng_lld.h (hunk 7 above)
Comment (networkfusion) [reply targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/rng_lld.h]: I would like to keep these 3 as one is used and the other 2 are options (although not currently used...

File: targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/rng_lld.h (hunk 10 above)
Comment (networkfusion) [reply targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/rng_lld.h]: done.

File: targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/src/stm32_rng/hal_stm32_rng.c (hunk 9 above)
Comment (networkfusion) [reply targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/src/stm32_rng/hal_stm32_rng.c]: done.

File: targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/rng_lld.h (hunk 8 above)
Comment (networkfusion) [reply targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/rng_lld.h]: done.

File: targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/rng_lld.h (hunk 6 above)
Comment (networkfusion) [reply targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/rng_lld.h]: see below...

File: targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/rng_lld.h (hunk 4 above)
Comment (networkfusion) [reply targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/rng_lld.h]: see below...

File: targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/README.md (hunk 2 above)
Comment (networkfusion) [reply targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/ports/STM32/LLD/RNGv1/README.md]: done..

File: targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/include/stm32_rng/hal_stm32_rng.h (hunk 1 above)
Comment (networkfusion) [reply targets/CMSIS-OS/ChibiOS/nf-overlay/os/hal/include/stm32_rng/hal_stm32_rng.h]: done...

**Pull request merged**